runtime/j9vm/asgct.cpp

```diff
@@ -112,19 +146,37 @@ protectedASGCT(J9PortLibrary *portLib, void *arg)
 				J9JITExceptionTable *metaData = jitConfig->jitGetExceptionTableFromPC(currentThread, rip);
 				if (NULL != metaData) {
 					greg_t rsp = regs[REG_RSP];
-					jitPushResolveFrame(currentThread, (UDATA*)rsp, (U_8*)rip);
+					/* Build a JIT resolve frame on the C stack to avoid writing to the java
+					 * stack in the signal handler. Update the J9VMThread roots to point to
+					 * the resolve frame (will be restored in the caller).
+					 */
+					J9SFJITResolveFrame resolveFrame = {
+						NULL, /* savedJITException */
+						J9_SSF_JIT_RESOLVE, /* specialFrameFlags */
+						0, /* parmCount */
+						(U_8*)rip, /* returnAddress */
+						(UDATA*)(((U_8 *)rsp) + J9SF_A0_INVISIBLE_TAG) /* taggedRegularReturnSP */
+					};
+					currentThread->pc = (U_8*)J9SF_FRAME_TYPE_JIT_RESOLVE;
+					currentThread->arg0EA = (UDATA*)&(resolveFrame.taggedRegularReturnSP);
+					currentThread->literals = NULL;
+					currentThread->sp = (UDATA*)&resolveFrame;
 				}
 			}
 		}
-		J9StackWalkState walkState;
+		/* Disable the JIT metadata cache. The cache may be in an inconsistent state, and may
+		 * need to allocate memory (which is undesireable in a signal handler).
+		 */
+		currentThread->jitArtifactSearchCache = (void*)((UDATA)currentThread->jitArtifactSearchCache | J9_STACKWALK_NO_JIT_CACHE);
+		/* Disable trace and assertions. The current thread may be in the process of updating the trace
+		 * buffers when it was interrupted. Assertions during ASGCT are also undesireable, better to let
+		 * the thread continue on to crash, which is handled.
+		 */
+		currentThread->privateFlags2 |= J9_PRIVATE_FLAGS2_ASYNC_GET_CALL_TRACE;
```

Comment:
@gacholio Should this flag be reset after the stack walk?

Reply:
It is reset in the caller, along with all the other modified values.